### PR TITLE
chore: Fix event emitters

### PR DESCRIPTION
**Pull request by kyle-ssg**
This fixes 2 issues

- https://github.com/Flagsmith/flagsmith-js-client/issues/214, useFlags was not updating because event listeners were removed
- If you instantiate the provider without options, useFlags will fail to update on changes. Prior to this PR, init setup event listeners via onChange. This is resolved by having onChange be part of the class rather than setup in flagsmith.init

How did I test this ?

- Tested with the project in https://github.com/Flagsmith/flagsmith-js-examples/pull/8, without this change logging in and out doesn't update useFlags.
- Tested on existing Next.JS example

### CODE REVIEW

File: flagsmith-core.ts [hunk 1]
```diff
@@ -131,11 +131,22 @@ const Flagsmith = class {
             })
     };
 
+    _onChange: IInitConfig['onChange'] = (previousFlags, params, loadingState) => {
```
Comment (kyle-ssg) [inline flagsmith-core.ts:134]: onChange was initialised on flagsmith.init, this means that triggers weren't setup if flagsmith.init wasn't called in an SSR scenario. Now this is done on construction.

File: emitter.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
+interface EventCallback {
```
Comment (kyle-ssg) [inline emitter.ts:1]: A simple event emitter where on/once return a function that removes exactly that listener.

File: react.tsx [hunk 3]
```diff
@@ -135,37 +135,37 @@ export function useFlags<F extends string=string, T extends string=string>(_flag
...
-    events.once('event', eventListener)
-    
     if (firstRender.current) {
         firstRender.current = false;
         flagsmith?.log("React - Initialising event listeners")
     }
 
     useEffect(()=>{
         return () => {
-            events.off('event', eventListener)
```
Comment (kyle-ssg) [inline react.tsx:156]: This was incorrectly removing multiple listeners, new approach only removes the ref's listener

**Review by novakzaballa — APPROVED** [review]
LGTM

**Pull request merged**